cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataEntity.java

```diff
@@ -26,7 +26,34 @@
 import javax.annotation.Nullable;
 
 /**
- * Entity representation for Metadata
+ * Entity representation for Metadata. This representation support representing any of the existing
+ * CDAP entities which support metadata or any custom CDAP resource/entity. MetadataEntity are just ordered
+ * <p>
+ * key-value pair which represent the resource/entity.
+ * <p>
+ * Example usage:
+ * Creating a MetadataEntity for stream
+ * <pre>
+ *  MetadataEntity metadataEntity = MetadataEntity.ofNamespace("myNamespace").append(MetadataEntity.STREAM, "myStream");
+ * </pre>
+ * <p>
+ * Creating a Metadata for a field in dataset which is a custom resource
+ * <pre>
+ * MetadataEntity metadataEntity = MetadataEntity.ofNamespace("myNamespace")
+ *   .append(MetadataEntity.DATASET, "myDataset").append("field", "myField");
+ * </pre>
+ * <p>
+ * By default the key of the last key-value pair becomes the type of the MetadataEntity but it is also possible to
+ * change the type on demand to represent the MetadataEntity to represent entities/resources whose type are not the key
+ * of last key value pair. An example of this is ApplicationId which ends with "version" although the type is
+ * application.
+ * <pre>
+ * MetadataEntity metadataEntity = MetadataEntity.ofNamespace("myNamespace")
+ *   .append(MetadataEntity.APPLICATION, "myApplication")
+ *   .append(MetadataEntity.VERSION, "appVersion");
+ * metadataEntity = metadataEntity.changeType(MetadataEntity.APPLICATION);
```

Comment:
I added a check for this in recent changes but your idea of appendAsType looks good as it's more specific. Will do this change.

Reply:
Although since append always return a new MetadataEntity and I think we should make appendAsType also return new MetadataEntity and allow user to keep changing the type. It will help in the following case: 
Say, I have a MetadataEntity representing a Dataset:
```
MetadataEntity md = new MetadataEntity.ofNamepspace("myNs").appendAsType(MetadataEntity.Dataset, "myDs")
```

Now being able to appendAsType again will allow me to do something like this

```
MetadataEntity anotherOne = md.appendAsType("k1", "v1").append("k2", "v2");
```